packages/http-server/src/server.ts

```diff
@@ -10,63 +11,75 @@ import { IPrismDiagnostic } from '@stoplight/prism-core';
+import { MicriHandler } from 'micri/dist';
+
+function searchParamsToNameValues(searchParams: URLSearchParams): IHttpNameValues {
+  const params = {};
+  for (const key of searchParams.keys()) {
+    const values = searchParams.getAll(key);
+    params[key] = values.length === 1 ? values[0] : values;
+  }
+  return params;
+}
+
+function addressInfoToString(address: AddressInfo | string | null) {
+  if (!address) return '';
+  const a = address as AddressInfo;
+  return `http://${a.address}:${a.port}`;
+}
+
+function parseRequestBody(request: IncomingMessage) {
+  // if no body provided then return null instead of empty string
+  if (
+    request.headers['content-type'] === undefined
+    && request.headers['transfer-encoding'] === undefined
+    && (request.headers['content-length'] === '0' || request.headers['content-length'] === undefined)
+  ) {
+    return null;
+  }
+
+  if (typeIs(request, ['application/json', 'application/*+json'])) {
+    return json(request);
+  } else {
+    return text(request);
+  }
+}
 
 export const createServer = (operations: IHttpOperation[], opts: IPrismHttpServerOpts): IPrismHttpServer => {
   const { components, config } = opts;
 
-  const server = fastify({
-    logger: components.logger,
-    disableRequestLogging: true,
-    modifyCoreObjects: false,
-  });
-
-  if (opts.cors) server.register(fastifyCors, { origin: true, credentials: true });
-
-  server.addContentTypeParser('*', { parseAs: 'string' }, (req, body, done) => {
-    if (typeIs(req, ['application/*+json'])) {
-      try {
-        return done(null, JSON.parse(body));
-      } catch (e) {
-        return done(e);
-      }
-    }
+  const micriHandler: MicriHandler = async (request: IncomingMessage, reply: ServerResponse) => {
+    const {
+      url,
+      method,
+      headers,
+      socket,
+    } = request;
 
-    if (typeIs(req, ['application/x-www-form-urlencoded'])) {
-      return done(null, body);
+    if (!url) {
+      return TE.left(new Error('Missing URL in request!'));
     }
 
-    const error: Error & { status?: number } = new Error(`Unsupported media type.`);
-    error.status = 415;
-    Error.captureStackTrace(error);
-    return done(error);
-  });
+    const body = await parseRequestBody(request);
 
-  const prism = createInstance(config, components);
-
-  const replyHandler: fastify.RequestHandler = (request, reply) => {
-    const {
-      req: { method, url },
-      body,
-      headers,
-      query,
-    } = request;
+    const { searchParams, pathname } = new URL(url, addressInfoToString(socket.address()));
 
     const input = {
       method: (method ? method.toLowerCase() : 'get') as HttpMethod,
       url: {
-        path: (url || '/').split('?')[0],
-        query,
-        baseUrl: query.__server,
+        path: pathname,
+        baseUrl: searchParams.get('__server') || undefined,
+        query: searchParamsToNameValues(searchParams),
       },
-      headers,
+      headers: headers as IHttpNameValue,
```

Comment:
I am afraid this cast is not legit, since the `IHttpNameValue` is just a string, while the IncomingHeaders *could* be an array, and we'll need to fix these.

Reply:
Good point. I need to dig into that, for now it seems that we don't support multiple headers with the same key inside Prism. While micri can provide such to us. The question is:
- shall we support multiple headers of the same name or
- override them in such case so it results with single string?

I will check how fastify was dealing with that.

Reply:
Fastify does not care about it: https://github.com/fastify/fastify/blob/master/fastify.d.ts#L34

Reply:
Ok, so Prism is wrong about headers; The duplicates are allowed in certain cases: https://nodejs.org/api/http.html#http_message_headers

Reply:
Looks like we've been supporting the happy-path where there is always a single string. For me, it seems like a bug and we should fix it in a separate task. @XVincentX agree?

Reply:
Hmm I think in reality Fastify has been grouping the headers for us, and we should replicate the same thing here.

<img width="942" alt="image" src="https://user-images.githubusercontent.com/1416224/72363799-bcc2ec80-36f5-11ea-843c-e6ad05fa95e3.png">

Reply:
Aha! It is not Fastify who is grouping the headers, it is node's http. See here: https://nodejs.org/api/http.html#http_message_headers. There is list of rules for headers, Set-Cooke has special treatment, Cookie is always an array and custom headers are grouped with comma as in your example.

Fastify was doing physically nothing to headers (https://github.com/fastify/fastify/blob/master/fastify.d.ts#L34). It just made typing loose (`{ [key: string]: any }`) so Prism was not complaining.

To sum this up, this PR **does not introduce any bug regarding headers treatment**. But we definitely had bug before and still have one.

I think about separate PR because due to different treatment of different headers it might be not so simple to tackle.

Reply:
Ok that's fair. Let's put a comment and we can handle that in a different PR.

Reply:
Allright. Let's put a comment and figure it out in a following PR.